src/core/cm/unitconfig/unitconfig.cpp

```diff
@@ -0,0 +1,244 @@
+
+namespace aos::cm::unitconfig {
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+Error UnitConfig::Init(const String& cfgFile, nodeinfoprovider::NodeInfoProviderItf& nodeInfoProvider,
+    NodeConfigHandlerItf& nodeConfigHandler, JSONProviderItf& jsonProvider)
+{
+    LOG_INF() << "Init unit config";
+
+    mCfgFile           = cfgFile;
+    mNodeInfoProvider  = &nodeInfoProvider;
+    mNodeConfigHandler = &nodeConfigHandler;
+    mJSONProvider      = &jsonProvider;
+
+    return LoadConfig();
+}
+
+Error UnitConfig::GetUnitConfigStatus(UnitConfigStatus& status)
+{
+    LockGuard lock {mMutex};
+
+    LOG_DBG() << "Get unit config status";
+
+    status.mVersion = mUnitConfig.mVersion;
+
+    if (!mUnitConfigError.IsNone()) {
+        status.mState = UnitConfigStateEnum::eFailed;
+        status.mError = mUnitConfigError;
+
+        return ErrorEnum::eNone;
+    }
+
+    status.mState = UnitConfigStateEnum::eInstalled;
+
+    return ErrorEnum::eNone;
+}
+
+Error UnitConfig::CheckUnitConfig(const aos::UnitConfig& config)
+{
+    LockGuard lock {mMutex};
+
+    LOG_DBG() << "Check unit config" << Log::Field("Version", config.mVersion);
+
+    if (!mUnitConfigError.IsNone() && !mUnitConfig.mVersion.IsEmpty()) {
+        LOG_WRN() << "Skip unit config version check due to error" << Log::Field(mUnitConfigError);
+    } else if (auto err = CheckVersion(config.mVersion); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    auto nodeIds = MakeUnique<StaticArray<StaticString<cIDLen>, cMaxNumNodes>>(&mAllocator);
+
+    if (auto err = mNodeInfoProvider->GetAllNodeIds(*nodeIds); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    for (const auto& id : *nodeIds) {
+        auto nodeConfigStatus = MakeUnique<NodeConfigStatus>(&mAllocator);
+
+        if (auto err = mNodeConfigHandler->GetNodeConfigStatus(id, *nodeConfigStatus); !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+
+        if (nodeConfigStatus->mVersion != config.mVersion || !nodeConfigStatus->mError.IsNone()) {
+            auto nodeConfig = MakeUnique<NodeConfig>(&mAllocator);
+
+            if (auto err = FindNodeConfig(id, "", config, *nodeConfig); !err.IsNone()) {
+                return err;
+            }
+
+            if (auto err = mNodeConfigHandler->CheckNodeConfig(id, *nodeConfig); !err.IsNone()) {
+                return AOS_ERROR_WRAP(err);
+            }
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error UnitConfig::GetNodeConfig(const String& nodeID, const String& nodeType, NodeConfig& config)
+{
+    LockGuard lock {mMutex};
+
+    LOG_DBG() << "Get node config" << Log::Field("NodeID", nodeID) << Log::Field("NodeType", nodeType);
+
+    return FindNodeConfig(nodeID, nodeType, mUnitConfig, config);
```

Comment:
I guess we shall return default node config if node config not found for specific  node.
Also we should return error if unit config in failed state.